x-pack/auditbeat/processors/sessionmd/processdb/names.go

```diff
@@ -0,0 +1,75 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build linux
+
+package processdb
+
+import (
+	"os/user"
+	"sync"
+)
+
+type cval struct {
+	name  string
+	found bool
+}
+
+type namesCache struct {
+	mutex  sync.RWMutex
+	users  map[string]cval
+	groups map[string]cval
+}
+
+// newNamesCache will return a new namesCache, which can be used to get mappings
+// of user and group IDs to names.
+func newNamesCache() *namesCache {
+	u := namesCache{
+		users:  make(map[string]cval),
+		groups: make(map[string]cval),
+	}
+	return &u
+}
+
+// getUserName will return the name associated with the user ID, if it exists
+func (u *namesCache) getUserName(id string) (string, bool) {
+	u.mutex.Lock()
+	defer u.mutex.Unlock()
+
+	val, ok := u.users[id]
+	if ok {
+		return val.name, val.found
+	}
+	user, err := user.LookupId(id)
+	cval := cval{}
+	if err != nil {
+		cval.name = ""
+		cval.found = false
+	} else {
+		cval.name = user.Username
+		cval.found = true
+	}
+	return cval.name, cval.found
```

Comment:
If I get this correctly before we return it we should save `cval` in the map right? That said, I think that the cache sure helps with the performance (especially if we are talking about LDAP-based users) but on the other hand it can affect the reliability of the data, e.g.

```
$ sudo useradd jane
$ sudo su jane
$ id
uid=1000(jane) gid=1001(jane) groups=1001(jane)
$ exit
$ sudo usermod -l new_jane jane
$ sudo su new_jane
$ id
uid=1000(new_jane) gid=1001(jane) groups=1001(jane)
```

I just renamed a user but the Id remains the same so better to search for it every-time?!?

Reply:
I've removed all the caching here, it turns out it has no benefit. I've also tried changing the username as the benchmarking ran, and it returned the new username.